pkg/virt-launcher/virtwrap/device/hostdevice/sriov/networkstatus-pcipool.go

```diff
@@ -0,0 +1,127 @@
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	v1 "kubevirt.io/api/core/v1"
+)
+
+type PCIAddressPoolWithNetworkStatus struct {
+	networkNameToPciAddress map[string]string
+}
+
+// NewPCIAddressPoolWithNetworkStatus creates a PCI address pool based on the network status multus annotation
+func NewPCIAddressPoolWithNetworkStatus(sriovIfaces []v1.Interface, multusNetworks []v1.Network, vmiAnnotations map[string]string) (*PCIAddressPoolWithNetworkStatus, error) {
+	pool := &PCIAddressPoolWithNetworkStatus{
+		networkNameToPciAddress: make(map[string]string),
+	}
+
+	if err := pool.loadNetworkNameToPciAddress(sriovIfaces, multusNetworks, vmiAnnotations); err != nil {
+		return nil, err
+	}
+
+	return pool, nil
+}
+
+func (p *PCIAddressPoolWithNetworkStatus) loadNetworkNameToPciAddress(sriovIfaces []v1.Interface, multusNetworks []v1.Network, vmiAnnotations map[string]string) error {
+	multusInterfaceNameToNetworkStatusMap, err := loadMultusInterfaceNameToNetworkStatus(vmiAnnotations)
+	if err != nil {
+		return err
+	}
+	sriovNetworkToMultusInterfaceNameMap := loadNetworkNameToMultusInterfaceName(sriovIfaces, multusNetworks)
+
+	for networkName, multusInterfaceName := range sriovNetworkToMultusInterfaceNameMap {
+		networkStatusEntry := multusInterfaceNameToNetworkStatusMap[multusInterfaceName]
+		if networkStatusEntry.DeviceInfo == nil || networkStatusEntry.DeviceInfo.Pci == nil {
+			return errors.New(fmt.Sprintf("SRIOV interface %s has no corresponding device-info/pci-address in the appropriate network status annotations", networkName))
+		}
+
+		pciAddress := networkStatusEntry.DeviceInfo.Pci.PciAddress
+		if pciAddress == "" {
+			return errors.New(fmt.Sprintf("failed to associate pci-address to SRIOV interface %s", networkName))
+		}
+		p.networkNameToPciAddress[networkName] = pciAddress
+	}
+	return nil
+}
+
+func (p *PCIAddressPoolWithNetworkStatus) Pop(networkName string) (string, error) {
+	pciAddress, exists := p.networkNameToPciAddress[networkName]
+	if !exists {
+		return "", fmt.Errorf("resource for SR-IOV network %s does not exist", networkName)
+	}
+
+	delete(p.networkNameToPciAddress, networkName)
+	return pciAddress, nil
+}
+
+func loadNetworkNameToMultusInterfaceName(interfaceList []v1.Interface, networkList []v1.Network) map[string]string {
+	networkToMultusInterfaceNameMap := map[string]string{}
+	for i, network := range networkList {
+		// multusInterfaceName follows a predefined indexation of the non-default multus networks (See pkg/virt-controller/services/multus_annotations.go: generateMultusCNIAnnotation.
+		// In order to not break this convention, this for loop should not break/continue
+		multusInterfaceName := fmt.Sprintf("net%d", i+1)
+		associatedNetwork := getIfaceByNetworkName(interfaceList, network.Name)
+		if associatedNetwork != nil {
+			networkToMultusInterfaceNameMap[network.Name] = multusInterfaceName
+		}
```

Comment:
It is not clear from the variables names that this logic only maps the network names to which SR-IOV interfaces are connected.

Reply:
Fixed